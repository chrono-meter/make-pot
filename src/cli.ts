#!/usr/bin/env node

import process from "node:process";
import * as yargs from "yargs";
import { hideBin } from "yargs/helpers";
import { getArgs } from "./cli/getArgs.js";
import { getJsonArgs } from "./cli/getJsonArgs.js";
import makeJsonCommand from "./jsonCommand.js";
import makepotCommand from "./potCommand.js";
import type { Args, MakeJsonArgs } from "./types.js";

/** Main execution */
// Get the selected command
const r = yargs
	.default(hideBin(process.argv))
	.options({
		makejson: {
			describe: "Make JSON file",
			type: "boolean",
			default: false,
		},
	})
	.parseSync() as { makejson: boolean };

// Execute the command
if (!r.makejson) {
	makepotCommand(getArgs() as Args);
} else {
	makeJsonCommand(getJsonArgs() as MakeJsonArgs);
}
