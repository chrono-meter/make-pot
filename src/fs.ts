import { type Args } from './types'
import path from 'path'
import fs from 'fs'

/**
 * Ensures that a folder exists at the specified path.
 *
 * @param {string | undefined} folderPath - The path of the folder to ensure existence for.
 * @return {string} - The path of the folder, or '.' if folderPath is undefined.
 */
function ensureFolderExists(folderPath: string | undefined): string {
	if (folderPath === undefined) {
		return '.'
	}
	try {
		// Check if the folder exists
		fs.accessSync(
			path.resolve(folderPath),
			fs.constants.R_OK | fs.constants.W_OK
		)
	} catch (error) {
		if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
			// The Folder does not exist, so create it
			fs.mkdirSync(folderPath, { recursive: true })
			console.log(`Folder created: ${folderPath}`)
			return folderPath
		}
	}
	return folderPath
}

/**
 * Writes the .pot file to disk
 * @param args
 * @param fileContent
 */
export async function writePotFile(args: Args, fileContent: string) {
	// the path to the .pot file
	const potFilePath = args.destination
		? path.join(process.cwd(), args.destination)
		: path.join(process.cwd())

	if (ensureFolderExists(potFilePath)) {
		fs.writeFileSync(
			path.join(potFilePath, `${args.slug}.pot`),
			fileContent
		)
	}
}
