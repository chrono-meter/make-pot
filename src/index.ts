#!/usr/bin/env node
import { makePot } from './makePot'

import { getArgs } from './cliArgs'

import yargs from 'yargs'

import { name, version } from '../package.json'

/** Main execution */
const args = getArgs()
if (Object.keys(args).length > 0) {
	console.log(name + ' version: ' + version)
	const timeStart = new Date()
	makePot(args)
		.then(() => {
			const timeEnd = new Date()
			console.log(
				`ðŸš€ Translation Pot file created in ${timeEnd.getTime() - timeStart.getTime()}ms`
			)
		})
		.catch((error) => {
			console.error(error)
		})
} else {
	yargs.showHelp()
}
