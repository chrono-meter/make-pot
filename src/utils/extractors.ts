import { Block, SetOfBlocks } from "gettext-merger";

/**
 * Returns the key of an object based on its value
 *
 * @param object the object that contains the key
 * @param value the key that we want to get
 * @return {Record<string, string>} the filtered keys
 */
export function getKeyByValue(
	object: Record<string, unknown>,
	value: string,
): string | undefined {
	return Object.keys(object).find((key) => object[key] === value) ?? undefined;
}

/**
 * Returns a gettext translation object
 *
 * @param label the label of the translation
 * @param string the string of the translation
 * @param filePath the file path of the translation
 */
export const buildBlock = (
	label: string,
	string: string,
	filePath: string[] | undefined = undefined,
): Block => {
	const block = new Block([]);
	block.msgctxt = undefined;
	block.msgid = string;
	block.msgid_plural = "";
	block.msgstr = [];
	block.comments = {};
	if (label) {
		block.comments.extracted = [label];
	}
	if (filePath?.length) {
		block.comments.reference = filePath;
	}
	return block;
};

/**
 * Extracts strings from parsed JSON data.
 *
 * @param {Record<string, any> | Parser.SyntaxNode} parsed - The parsed JSON data or syntax node.
 * @param {string | Parser} filename - The filename or parser.
 * @param filepath - the path to the file being parsed
 * @return {SetOfBlocks} An array of translation strings.
 */
export function yieldParsedData(
	parsed: Block[],
	filename: "block.json" | "theme.json",
	filepath: string,
): SetOfBlocks {
	const gettextTranslations: SetOfBlocks = new SetOfBlocks([], filepath);

	if (parsed.length === 0) {
		return gettextTranslations;
	}

	// set the path of the translation
	gettextTranslations.path = filepath;

	for (const item of parsed) {
		const block = buildBlock(
			item.msgid,
			item.msgctxt as string,
			item.comments?.reference,
		);

		if (block) {
			gettextTranslations.blocks.push(block);
		}
	}

	return gettextTranslations;
}
