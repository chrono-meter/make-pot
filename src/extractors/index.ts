import * as path from 'path'
import { readFileAsync } from '../fs'
import { parseJsonCallback } from './json'
import { doTree } from '../parser/tree'
import { SetOfBlocks } from 'gettext-merger'
import { allowedFiles } from '../const'

/**
 * Parse a file and extract strings asynchronously
 *
 * @return {Promise<SetOfBlocks>}
 */
export async function parseFile(
	file: string,
	filePath: string = ''
): Promise<SetOfBlocks | undefined> {
	const filename = path.basename(file)
	const ext = path.extname(file).replace(/^./, '')
	const fileRealPath = path.resolve(filePath, file)

	if (ext === 'json') {
		if (filename === 'theme.json' || filename === 'block.json') {
			return readFileAsync(fileRealPath).then(async (sourceCode) => {
				return await parseJsonCallback(sourceCode, filePath, filename)
			})
		}
	}

	if (allowedFiles.includes(ext)) {
		// read the file
		return readFileAsync(fileRealPath).then((sourceCode) => {
			return doTree(sourceCode, file)
		})
	}
}
