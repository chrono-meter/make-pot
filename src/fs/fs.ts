import fs, { writeFileSync } from "node:fs";
import { readFile } from "node:fs/promises";
import path from "node:path";
import type { Args } from "../types";

/**
 * Ensures that a folder exists at the specified path.
 *
 * @param {string | undefined} folderPath - The path of the folder to ensure existence for.
 * @return {string} - The path of the folder, or '.' if folderPath is undefined.
 */
function ensureFolderExists(folderPath: string | undefined): string {
	if (folderPath === undefined) {
		return ".";
	}
	try {
		// Check if the folder exists
		fs.accessSync(
			path.resolve(folderPath),
			fs.constants.R_OK | fs.constants.W_OK,
		);
	} catch (error) {
		if ((error as NodeJS.ErrnoException).code === "ENOENT") {
			// The Folder does not exist, so create it
			fs.mkdirSync(folderPath, { recursive: true });
			console.log(`Folder created: ${folderPath}`);
			return folderPath;
		}
	}
	return folderPath;
}

/**
 * Gets the charset of the .pot file
 *
 * @param charset the charset of the .pot file
 * @return the charset of the .pot file
 */
export function getCharset(charset: string | undefined): BufferEncoding {
	if (!charset) {
		return "latin1";
	}
	// we need to check if the charset is valid otherwise we return latin1 that is a common alias for ISO-8859-1 and the default charset for pot files
	switch (charset.toLowerCase()) {
		case "utf-8":
			return "utf-8";
		default:
			return "latin1";
	}
}

export function getEncodingCharset(charset: string | undefined): string {
	if (!charset) {
		return "iso-8859-1";
	}
	// we need to check if the charset is valid otherwise we return utf-8 that is a common alias for ISO-8859-1 and the default charset for pot files
	switch (charset.toLowerCase()) {
		case "utf-8":
			return "utf-8";
		default:
			return "iso-8859-1";
	}
}

/**
 * The output path for the pot file.
 * @param outpath - the output path for the pot/json files
 * @return {string} - the output path
 */
export function getOutputPath(outpath?: string): string {
	return path.join(process.cwd(), outpath ?? "languages");
}

/**
 * The output path for the pot file.
 * @param args - the command line arguments
 */
function getOutputFilePath(args: Args): string {
	const { out, headers, options } = args.paths;
	let i18nFolder = out ?? headers?.domainPath ?? "languages";

	// Remove leading and trailing slashes
	i18nFolder = i18nFolder.replace(/^\/+|\/+$/g, "");

	const extension = options?.json ? "json" : "pot";

	return path.join(process.cwd(), i18nFolder, `${args.slug}.${extension}`);
}

/**
 * Writes the .pot file to disk
 *
 * @param fileContent the content of the .pot file
 * @param args the command line arguments passed to the program
 */
export function writeFile(fileContent: string, args: Args): void {
	const dest = getOutputFilePath(args);

	if (ensureFolderExists(path.dirname(dest))) {
		// get the encoding charset
		const encodingCharset = getCharset(args.options?.charset);
		console.log(`File created at ${dest}`);

		// write the file
		const potBuffer = Buffer.from(fileContent);
		writeFileSync(dest, potBuffer.toString(encodingCharset), {
			encoding: encodingCharset,
		});
	} else {
		console.log(`Folder ${dest} does not exist and cannot be created`);
	}
}

/**
 * The async version of fs.readFile method
 * @param path the path of the file to read
 * @return the content of the file as a string
 */
export async function readFileAsync(path: string): Promise<string> {
	return readFile(path, "utf-8");
}
