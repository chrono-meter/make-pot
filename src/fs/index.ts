import * as path from 'path'
import fs from 'node:fs'

/**
 * Ensures that a folder exists at the specified path.
 *
 * @param {string | undefined} folderPath - The path of the folder to ensure existence for.
 * @return {string} - The path of the folder, or '.' if folderPath is undefined.
 */
function ensureFolderExists(folderPath: string | undefined): string {
	if (folderPath === undefined) {
		return '.'
	}
	try {
		// Check if the folder exists
		fs.accessSync(
			path.resolve(folderPath),
			fs.constants.R_OK | fs.constants.W_OK
		)
	} catch (error) {
		if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
			// The Folder does not exist, so create it
			fs.mkdirSync(folderPath, { recursive: true })
			console.log(`Folder created: ${folderPath}`)
			return folderPath
		}
	}
	return folderPath
}

/**
 * Writes the .pot file to disk
 *
 * @param fileContent the content of the .pot file
 * @param dest the path of the .pot file to write
 */
export function writeFile(fileContent: string, dest: string) {
	if (ensureFolderExists(path.dirname(dest))) {
		fs.writeFileSync(dest, fileContent)
	}
}

export async function readFileAsync(path: string): Promise<string> {
	return fs.promises.readFile(path, 'utf-8')
}
