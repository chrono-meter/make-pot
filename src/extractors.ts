import path from 'path'
import fs, { readFileSync } from 'fs'
import { type Args, type TranslationString } from './types'
import { pkgJsonHeaders } from './const'
import { getCommentBlock } from './utils'
import Parser from 'tree-sitter'
import type { SingleBar } from 'cli-progress'
import { jsonString, parseJsonFile } from './extractors-json'
import { parsePHPFile } from './extractors-php'
import { extractStrings } from './tree'
import { extractFileData } from './extractors-text'

/**
 * Extracts the names from an array of items.
 *
 * @param {unknown[]} items - The array of items.
 * @return {string[]} An array of names extracted from the items.
 */
export const extractNames = (items: { name: string }[]): string[] => items.map((item) => item.name)

/**
 * Extracts strings from parsed JSON data.
 *
 * @param {Record<string, any> | Parser.SyntaxNode} parsed - The parsed JSON data or syntax node.
 * @param {string | Parser} filename - The filename or parser.
 * @param opts - The metadata of this translation string.
 * @return {TranslationString[]} An array of translation strings.
 */
export function yieldParsedData(
	parsed: Record<string, string | string[]> | Parser.SyntaxNode,
	filename: string | Parser,
	opts: { filepath: string; stats?: { bar: SingleBar } }
): Promise<TranslationString[]> {
	return new Promise<TranslationString[]>((resolve) =>
		resolve(
			Object.entries(parsed as Record<string, string | string[]>)

				// return the translations for each key in the json data
				.map(([key, jsonData]) => {
					// if is a string return a single json string
					if (typeof jsonData === 'string') {
						return jsonString(
							key,
							jsonData,
							opts.filepath,
							filename as 'block.json' | 'theme.json'
						)
					} else {
						opts.stats?.bar.increment(0, { filename: 'not a string' })
					}

					if (!jsonData) {
						opts.stats?.bar.increment(0, {
							filename: `Skipping ${key} in ${opts.filepath} as ${filename} ... cannot parse data`,
						})

						return null
					}

					// if is an object return an array of json strings
					Object.entries(jsonData).map(([k, v]) => jsonString(k, v, opts.filepath))
				})
				.flat() as TranslationString[]
		)
	)
}

/**
 * Parses the source code using the specified language parser and extracts the strings from the file.
 *
 * @param {string} sourceCode - The source code to be parsed.
 * @param {Args} language - The language to be used for parsing.
 * @param {string} filepath - The path to the file being parsed.
 * @return {TranslationString[]} An array of translation strings.
 */
export function doTree(sourceCode: string, language: Parser, filepath: string) {
	// set up the parser
	const parser = new Parser()
	parser.setLanguage(language)

	// parse the file
	const tree = parser.parse(sourceCode) // Assuming parse is an async operation

	// extract the strings from the file and return them
	return extractStrings(tree.rootNode, language, filepath)
}

/**
 * Parse a file and extract strings asynchronously
 *
 * @param {object} args
 * @param {string} args.filepath - Path to the file to parse
 * @param {Parser|null} args.language - Language of the file to parse
 * @return {Promise<TranslationString[]>}
 */
export async function parseFile(args: {
	filepath: string
	language: Parser | string
}): Promise<TranslationString[] | null> {
	// check if the language is supported
	if (typeof args.language === 'string') {
		if (args.language === 'json') {
			return parseJsonFile(args) || []
		}
		console.log(`Skipping ${args.filepath}... No parser found for ${args.language} file`)
		return null
	}

	// read the file
	const sourceCode = readFileSync(path.resolve(args.filepath), 'utf8')

	// set up the parser
	const parser = new Parser()
	parser.setLanguage(args.language)

	// parse the file
	const tree = parser.parse(sourceCode) // Assuming parse is an async operation

	// extract the strings from the file and return them
	return extractStrings(tree.rootNode, args.language, args.filepath)
}

/**
 * Extracts package data from the given arguments and returns a record
 * containing the specified fields from the package.json file.
 *
 * @param {Args} args - The arguments for extracting package data.
 * @param {Record<string, string>} fields - The fields to extract from the package.json file. Default is pkgJsonHeaders.
 * @return {Record<string, string>} - A record containing the extracted package data.
 */
export function extractPackageData(args: Args, fields = pkgJsonHeaders): Record<string, string> {
	// TODO: package.json "files" could be used to get the file list
	const pkgJsonMeta: Record<string, string> = {}
	// read the package.json file
	const packageJsonPath = args.sourceDirectory
		? path.join(args.sourceDirectory, 'package.json')
		: 'package.json'
	if (fs.existsSync(packageJsonPath)) {
		const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'))
		// extract the fields from the package.json file
		for (const field of Object.keys(fields)) {
			if (packageJson[field] !== undefined) {
				pkgJsonMeta[field] = packageJson[field]
			}
		}
	}
	return pkgJsonMeta
}

/**
 * Extracts main file data based on the given arguments.
 *
 * @param {Args} args - The arguments for extracting the main file data.
 * @return {Record<string, string>} The extracted main file data.
 */
export function extractMainFileData(args: Args) {
	let fileData: Record<string, string> = {}
	const sourceDir = args.sourceDirectory
		? path.join(process.cwd(), args.sourceDirectory)
		: process.cwd()

	if (['plugin', 'block', 'generic'].includes(args.domain)) {
		const folderPhpFile = path.join(sourceDir, `${args.slug}.php`)

		if (fs.existsSync(folderPhpFile)) {
			const fileContent = fs.readFileSync(folderPhpFile, 'utf8')
			fileData = parsePHPFile(fileContent)

			if ('Plugin Name' in fileData) {
				console.log('Plugin file detected.')
				console.log(`Plugin file: ${folderPhpFile}`)
				args.domain = 'plugin'
			}
		} else {
			console.log('Plugin file not found.')
			console.log(`Missing Plugin filename: ${folderPhpFile}`)
		}
	} else if (['theme', 'theme-block'].includes(args.domain)) {
		const styleCssFile = path.join(sourceDir, 'style.css')

		if (fs.existsSync(styleCssFile)) {
			const fileContent = fs.readFileSync(styleCssFile, 'utf8')
			const commentBlock = getCommentBlock(fileContent)
			fileData = extractFileData(commentBlock)

			console.log('Theme stylesheet detected.')
			console.log(`Theme stylesheet: ${styleCssFile}`)
			args.domain = 'theme'
		} else {
			console.log(`Theme stylesheet not found in ${path.resolve(sourceDir)}`)
		}
	}

	return fileData
}
