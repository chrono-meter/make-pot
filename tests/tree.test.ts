import { describe, expect } from '@jest/globals'
import { doTree } from '../src/extractors'
import fs from 'fs'
import path from 'path'
// @ts-expect-error unused
import Js from 'tree-sitter-javascript'
// @ts-expect-error unused
import Php from 'tree-sitter-php'
// @ts-expect-error unused
import Ts from 'tree-sitter-typescript'

describe('doTree', () => {
	it('Should build pot file', async () => {
		const fileParsed = doTree(
			fs.readFileSync(
				path.join(__dirname, 'fixtures/sourcedir/file.php'),
				'utf8'
			),
			Php,
			'sourcedir/file.php'
		)
		console.log(fileParsed)
		expect(fileParsed).toMatchSnapshot()
	})
	it('Should parse js', async () => {
		const fileParsed = doTree(
			fs.readFileSync(
				path.join(__dirname, 'fixtures/block/javascript.js'),
				'utf8'
			),
			Js,
			'block/javascript.js'
		)
		console.log(fileParsed)
		expect(fileParsed).toMatchSnapshot()
	})
	it('Should parse TSX file and extract strings', async () => {
		const fileParsed = doTree(
			fs.readFileSync(
				path.join(__dirname, 'fixtures/block/SvgControls.tsx'),
				'utf8'
			),
			Ts.tsx,
			'SvgControls.tsx'
		)
		console.log(fileParsed)
		expect(fileParsed).toMatchSnapshot()
	})
})
