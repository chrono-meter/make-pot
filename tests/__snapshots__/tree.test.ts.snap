// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doTree Should build pot file 1`] = `
{
  "": {
    "Huge": {
      "comments": {
        "reference": "sourcedir/file.php:252",
      },
      "msgid": "Huge",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "Large": {
      "comments": {
        "reference": "sourcedir/file.php:247",
      },
      "msgid": "Large",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "Light green cyan to vivid green cyan": {
      "comments": {
        "reference": "sourcedir/file.php:189",
      },
      "msgid": "Light green cyan to vivid green cyan",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "Luminous vivid amber to luminous vivid orange": {
      "comments": {
        "reference": "sourcedir/file.php:194",
      },
      "msgid": "Luminous vivid amber to luminous vivid orange",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "Luminous vivid orange to vivid red": {
      "comments": {
        "reference": "sourcedir/file.php:199",
      },
      "msgid": "Luminous vivid orange to vivid red",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "Regular": {
      "comments": {
        "reference": "sourcedir/file.php:242",
      },
      "msgid": "Regular",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "Small": {
      "comments": {
        "reference": "sourcedir/file.php:237",
      },
      "msgid": "Small",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "Vivid cyan blue to vivid purple": {
      "comments": {
        "reference": "sourcedir/file.php:179",
      },
      "msgid": "Vivid cyan blue to vivid purple",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "Vivid green cyan to vivid cyan blue": {
      "comments": {
        "reference": "sourcedir/file.php:184",
      },
      "msgid": "Vivid green cyan to vivid cyan blue",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "light grayish magenta": {
      "comments": {
        "reference": "sourcedir/file.php:123",
      },
      "msgid": "light grayish magenta",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "sdasdasdasd": {
      "comments": {
        "reference": "sourcedir/file.php:7",
        "translator": "ciao2!",
      },
      "msgid": "sdasdasdasd",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "strong magenta": {
      "comments": {
        "reference": "sourcedir/file.php:118",
      },
      "msgid": "strong magenta",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "very dark gray": {
      "comments": {
        "reference": "sourcedir/file.php:133",
      },
      "msgid": "very dark gray",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "very light gray": {
      "comments": {
        "reference": "sourcedir/file.php:128",
      },
      "msgid": "very light gray",
      "msgid_plural": undefined,
      "msgstr": [],
    },
  },
}
`;

exports[`doTree Should build pot file 2`] = `
{
  "": {
    "%d thing.": {
      "comments": {
        "reference": "sourcedir/file.php:42",
      },
      "msgid": "%d thing.",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "%s comment": {
      "comments": {
        "reference": "sourcedir/file.php:30",
      },
      "msgid": "%s comment",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "%s comment.": {
      "comments": {
        "reference": "sourcedir/file.php:46",
      },
      "msgid": "%s comment.",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    ", and your zip code is ": {
      "comments": {
        "reference": "sourcedir/file.php:86",
      },
      "msgid": ", and your zip code is ",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "<strong>Version %1$s</strong> addressed %2$s bug.": {
      "comments": {
        "reference": "sourcedir/file.php:71",
        "translator": "",
      },
      "msgid": "<strong>Version %1$s</strong> addressed %2$s bug.",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "Blog Options": {
      "comments": {
        "reference": "sourcedir/file.php:3",
      },
      "msgid": "Blog Options",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "Last thing!": {
      "comments": {
        "reference": "sourcedir/file.php:40",
      },
      "msgid": "Last thing!",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "Post": {
      "comments": {
        "reference": "sourcedir/file.php:63",
      },
      "msgid": "Post",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "Posts:": {
      "comments": {
        "reference": "sourcedir/file.php:89",
      },
      "msgid": "Posts:",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "Search results for: %s": {
      "comments": {
        "reference": "sourcedir/file.php:78",
      },
      "msgid": "Search results for: %s",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "WordPress is the best!": {
      "comments": {
        "reference": "sourcedir/file.php:7",
      },
      "msgid": "WordPress is the best!",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "Your city is ": {
      "comments": {
        "reference": "sourcedir/file.php:86",
      },
      "msgid": "Your city is ",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "Your city is $city.": {
      "comments": {
        "reference": "sourcedir/file.php:74",
      },
      "msgid": "Your city is $city.",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "Your city is %1$s, and your zip code is %2$s.": {
      "comments": {
        "reference": "sourcedir/file.php:83",
      },
      "msgid": "Your city is %1$s, and your zip code is %2$s.",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "Your city is %s.": {
      "comments": {
        "reference": "sourcedir/file.php:11",
        "translator": "",
      },
      "msgid": "Your city is %s.",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "Your zip code is %2$s, and your city is %1$s.": {
      "comments": {
        "reference": "sourcedir/file.php:24",
        "translator": "",
      },
      "msgid": "Your zip code is %2$s, and your city is %1$s.",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "g:i:s a": {
      "comments": {
        "reference": "sourcedir/file.php:66",
        "translator": "",
      },
      "msgid": "g:i:s a",
      "msgid_plural": undefined,
      "msgstr": [],
    },
  },
}
`;

exports[`doTree Should parse TSX file and extract strings 1`] = `
{
  "": {
    "Edit Link": {
      "comments": {
        "reference": "SvgControls.tsx:93",
      },
      "msgid": "Edit Link",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "Replace SVG": {
      "comments": {
        "reference": "SvgControls.tsx:107",
      },
      "msgid": "Replace SVG",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "Unlink": {
      "comments": {
        "reference": "SvgControls.tsx:99",
      },
      "msgid": "Unlink",
      "msgid_plural": undefined,
      "msgstr": [],
    },
  },
}
`;

exports[`doTree Should parse js 1`] = `
{
  "": {
    "Hello World": {
      "comments": {
        "reference": "block/javascript.js:19",
      },
      "msgid": "Hello World",
      "msgid_plural": undefined,
      "msgstr": [],
    },
    "Simple Block": {
      "comments": {
        "reference": "block/javascript.js:7",
      },
      "msgid": "Simple Block",
      "msgid_plural": undefined,
      "msgstr": [],
    },
  },
}
`;
